lib/ResolverFactory.js

```diff
@@ -280,7 +290,8 @@ exports.createResolver = function(options) {
 		aliasFields.forEach(item => {
 			plugins.push(new AliasFieldPlugin("file", item, "resolve"));
 		});
-		if (symlinks) plugins.push(new SymlinkPlugin("file", "relative"));
+		if (symlinks && !pnpApi)
```

Comment:
Why is this necessary?

Reply:
Because ambiguities with `require.resolve`, whose return value is used both as input for `readFile` and `require`. This property means that the return value of `require.resolve` must both be a valid filesystem path, and a unique identifier for `require` call.

It doesn't cause issues in general cases, except when using peer dependencies - in this case you cannot uniquely identify a package via its location on the disk, because it might be instantiated multiple times in the dependency tree (each time with a different set of dependencies). In order to solve this, Yarn generates "virtual packages" for all packages that list peer dependencies, which effectively are symlinks to the cache. The symlink doesn't do anything, except that it gives semantic information used in the resolution (for example `webpack-cli-abc` would be registered with `webpack@1.0.0` in its dependencies, while `webpack-cli-123` would be registered with `webpack@2.0.0` - even though both would use the same symlink target of `/cache/webpack-cli-1.0.0`).

In order to work properly, this behavior requires the symlinks to be preserved, otherwise we lose the information previously stored.

I'll add a comment linking to this comment.

Reply:
hmm I understand the issue, but this need to be solved in some other way. Currently this change alters symlink behavior for all paths even non-pnp paths (i. e. local paths). It's ok to disable symlinks, but only for pnp-resolved paths.

You can do this i. e. by adding a flag to the `request` (i. e. `keepSymlinks`) when resolved by the pnp plugin and handle this flag in the SymlinkPlugin.